### PR TITLE
OVN: use config file via ConfigMap rather than environment variables

### PR DESCRIPTION
**Pull request by pecameron**
The ovn-config configMap can be mounted into the container and
values taken from there rather than using the environment variables.

The environment variables are preserved for existing users.

At some point (before this is merged) the ovnkube config file will be
used to configure ovnkube as well.

SDN-456
https://jira.coreos.com/browse/SDN-456

Signed-off-by: Phil Cameron <pcameron@redhat.com>

### CODE REVIEW

**Comment by pecameron** [comment]
@dcbw this has most of the needed changes. PTAL

File: dist/templates/ovn-setup.yaml.j2 [hunk 1]
```diff
@@ -130,3 +130,37 @@ data:
...
+  namespace: ovn-kubernetes
+data:
+  config:   |-
+    [default]
+    mtu="1400"
+
+    [kubernetes]
+    service-cidr="{{ svc_cidr | default('172.30.0.0/16') }}"
+    cluster-subnets="{{ net_cidr | default('10.128.0.0/14/23') }}"
+    ovn-config-namespace="ovn-kubernetes"
```
Comment (dcbw) [inline dist/templates/ovn-setup.yaml.j2]: @pecameron the default is already "ovn-kubernetes" so I think we can leave ovn-config-namespace out.

**Comment by pecameron** [comment]
@dcbw @squeed @girishmg This is now working. PTAL

File: dist/images/README_IMAGE.md [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+```
+At present version "4" and "3" are supported.
+
+# Version 4 interface
+
+## ovn-config configMap
+
+The ovn-config configMap is mounted into the container in directory
+```
+/var/run/ovn-config
```
Comment (danwinship) [inline dist/images/README_IMAGE.md]: For the `ovn-config` ConfigMap, I don't think there's any advantage to doing it this way rather than sticking with environment variables as in v3. Especially since `ovnkube.sh` still needs to have the code to deal with the environment variables anyway...

File: dist/images/README_IMAGE.md [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+In the yaml files each container has the volumeMounts: and the pod volumes: has configMap:
+```
+containers:
+- name:
+  volumeMounts:
+  - mountPath: /var/run/ovnkube-config
+    name: ovnkube-config-mount
+volumes:
+- configMap:
+    defaultMode: 420
```
Comment (danwinship) [inline dist/images/README_IMAGE.md]: Ugh. That's terrible, why haven't they fixed that API? Anyway, it looks like `0644` (aka decimal 420) is the default value anyway so you can just drop this.

File: dist/images/README_IMAGE.md [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    items:
+    - key: config
+      path: ovn_k8s.conf
```
Comment (danwinship) [inline dist/images/README_IMAGE.md]: If you either rename the key to `ovn_k8s.conf` in the ConfigMap, or else change `ovnkube.sh` to expect `/var/run/ovnkube-config/config` instead of `.../ovn_k8s.conf`, then you wouldn't need the `items` section here. (Likewise, if you keep the `ovn-config` configmap, you don't need the `items` section when mounting that at all, since you're mounting every item, and not renaming any of them.)

File: dist/images/README_IMAGE.md [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    name: ovnkube-config
+  name: ovnkube-config-mount
+```
+The following configmap contains the config file for onvkube:
+
+```
+apiVersion: v1
+data:
```
Comment (danwinship) [inline dist/images/README_IMAGE.md]: need `kind` and `metadata`

File: dist/images/README_IMAGE.md [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+  config: |-
+    [default]
+    mtu="1400"
+    cluster-subnets="10.128.0.0/14/23"
+
+    [kubernetes]
+    service-cidr="172.30.0.0/16"
+    ovn-config-namespace="openshift-ovn-kubernetes"
```
Comment (danwinship) [inline dist/images/README_IMAGE.md]: "ovn-kubernetes"

File: dist/images/README_IMAGE.md [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+volumes:
+- configMap:
+    defaultMode: 420
+    items:
+    - key: config
+      path: ovn_k8s.conf
+    name: ovnkube-config
+  name: ovnkube-config-mount
+```
+The following configmap contains the config file for onvkube:
```
Comment (danwinship) [inline dist/images/README_IMAGE.md]: typo "onvkube"

also, it's confusing to the reader that none of the values in this configmap match the values in the ovn-config configmap

File: dist/images/ovnkube.sh [hunk 8]
```diff
@@ -3,28 +3,28 @@
 #set -euo pipefail
 
 # This script is the entrypoint to the image.
-# Supports version 3 daemonsets
+# Supports version 3, and 4 daemonsets
 #    $1 is the daemon to start.
-#        In version 3 each process has a separate container. Some daemons start
+#        In versions 3 and 4 each process has a separate container. Some daemons start
```
Comment (danwinship) [inline dist/images/ovnkube.sh]: "Each process has a separate container". You don't need to specify "In versions 3 and 4" since there aren't any other versions any more

File: dist/images/ovnkube.sh [hunk 9]
```diff
@@ -33,27 +33,36 @@
...
 # OVN_GATEWAY_OPTS - the options for the ovn gateway
 # OVNKUBE_LOGLEVEL - log level for ovnkube (0..5, default 4) - v3
-# OVN_LOG_NORTHD - log level (ovn-ctl default: -vconsole:emer -vsyslog:err -vfile:info) - v3
-# OVN_LOG_NB - log level (ovn-ctl default: -vconsole:off -vfile:info) - v3
-# OVN_LOG_SB - log level (ovn-ctl default: -vconsole:off -vfile:info) - v3
-# OVN_LOG_CONTROLLER - log level (ovn-ctl default: -vconsole:off -vfile:info) - v3
+# OVN_LOG_NORTHD - log level (ovn-ctl default: -vconsole:emer -vsyslog:err -vfile:info) - v3, v4
+# OVN_LOG_NB - log level (ovn-ctl default: -vconsole:off -vfile:info) - v3, v4
+# OVN_LOG_SB - log level (ovn-ctl default: -vconsole:off -vfile:info) - v3, v4
+# OVN_LOG_CONTROLLER - log level (ovn-ctl default: -vconsole:off -vfile:info) - v3, v4
```
Comment (danwinship) [inline dist/images/ovnkube.sh]: The README said these were v3 only?

File: dist/images/ovnkube.sh [hunk 10]
```diff
@@ -85,11 +89,13 @@ ovs_options=${OVS_OPTIONS:-""}
...
 then
   k8s_token=$(cat /var/run/secrets/kubernetes.io/serviceaccount/token)
+  echo "k8s_token from file"
 else
   k8s_token=${K8S_TOKEN}
 fi
 
-K8S_CACERT=${K8S_CACERT:-/var/run/secrets/kubernetes.io/serviceaccount/ca.crt}
+k8s_cacert=${K8S_CACERT:-/var/run/secrets/kubernetes.io/serviceaccount/ca.crt}
+echo "k8s_cacert ${k8s_cacert}   K8S_CACERT ${K8S_CACERT}"
```
Comment (danwinship) [inline dist/images/ovnkube.sh]: leftover debugging? (likewise above `echo "k8s_token from file"`)

File: dist/images/ovnkube.sh [hunk 11]
```diff
@@ -100,15 +106,26 @@ ovn_controller_opts=${OVN_CONTROLLER_OPTS:-""}
 # set the log level for ovnkube
 ovnkube_loglevel=${OVNKUBE_LOGLEVEL:-4}
 
-# by default it is going to be a shared gateway mode, however this can be overridden to any of the other
-# two gateway modes that we support using `images/daemonset.sh` tool
-ovn_gateway_mode=${OVN_GATEWAY_MODE:-"shared"}
+# by default it is going to be a shared gateway mode, however this can be overridden to any
+# of the other two gateway modes that we support using `images/daemonset.sh` tool
+# v3
+ovn_gateway_mode=${OVN_GATEWAY_MODE}
```
Comment (danwinship) [inline dist/images/ovnkube.sh]: it looks like you removed the default here

File: dist/images/ovnkube.sh [hunk 12]
```diff
@@ -117,7 +134,7 @@ ovn_db_host=""
 # =========================================
 
 setup_ovs_permissions () {
-  if [ ${ovs_user_id:-XX} != "XX" ]; then
+  if [[ ${ovs_user_id:-XX} != "XX" ]] ; then
```
Comment (danwinship) [inline dist/images/ovnkube.sh]: if you're going to do style fixes on this:

    if [[ -n "${ovs_user_id}" ]]; then

(and again below)

File: dist/images/ovnkube.sh [hunk 13]
```diff
@@ -487,13 +520,13 @@ cleanup-ovs-server () {
...
   # delete any endpoint by name ovnkube-db
-  kubectl --server=${K8S_APISERVER} --token=${k8s_token} --certificate-authority=${K8S_CACERT} \
+  kubectl --server=${k8s_apiserver} --token=${k8s_token} --certificate-authority=${k8s_cacert} \
     delete ep -n ${ovn_kubernetes_namespace} ovnkube-db 2>/dev/null
 
   # create a new endpoint for the headless onvkube-db service without selectors
   # using the current host has the endpoint IP
-  ovn_db_host=$(getent ahosts $(hostname) | head -1 | awk '{ print $1 }')
-  kubectl --server=${K8S_APISERVER} --token=${k8s_token} --certificate-authority=${K8S_CACERT} create -f - << EOF
+  ovn_db_host=$(host ${K8S_NODE} | awk '{ print $4}')
```
Comment (danwinship) [inline dist/images/ovnkube.sh]: (Do we know that `host` is available with the set of packages installed in the ubuntu build? On Fedora it's in `bind-utils`, which we do appear to be installing from the Dockerfiles.)

File: dist/images/ovnkube.sh [hunk 14]
```diff
@@ -642,30 +676,35 @@ ovn-master () {
   # wait for northd to start
   wait_for_event process_ready ovn-northd
 
-  echo "=============== ovn-master (wait for ovn-nbctl daemon) ========== MASTER ONLY"
-  wait_for_event process_ready ovn-nbctl
```
Comment (danwinship) [inline dist/images/ovnkube.sh]: I don't think you meant to remove this. (Bad git rebase)

File: dist/images/ovnkube.sh [hunk 15]
```diff
@@ -698,7 +736,6 @@ ovn-controller () {
   wait_for_event process_ready ovn-controller
   echo "=============== ovn-controller ========== running"
 
-  sleep 4
```
Comment (danwinship) [inline dist/images/ovnkube.sh]: why was this there before and why is it ok to remove now?

File: dist/images/ovnkube.sh [hunk 16]
```diff
@@ -726,22 +763,37 @@ ovn-node () {
   wait_for_event process_ready ovn-controller
   sleep 1
 
-  echo "=============== ovn-node   --init-node"
+  echo "=============== ovn-node   --init-node  ${K8S_NODE}"
+  # TEMP HACK - WORKAROUND
+  # --gateway-mode=local works around a problem that
+  # results in loss of network connectivity when docker is
+  # restarted or ovs daemonset is deleted.
+  # TEMP HACK - WORKAROUND
```
Comment (danwinship) [inline dist/images/ovnkube.sh]: not sure what this is about...

File: dist/templates/ovn-setup.yaml.j2 [hunk 17]
```diff
@@ -130,3 +130,27 @@ data:
...
+    [kubernetes]
+    service-cidr="{{ svc_cidr | default('172.30.0.0/16') }}"
+    cluster-subnets="{{ net_cidr | default('10.128.0.0/14/23') }}"
+    apiserver="{{ k8s_apiserver.stdout }}"
+
+    [logging]
+    loglevel=4
+
+    [gateway]
+    mode=local
```
Comment (danwinship) [inline dist/templates/ovn-setup.yaml.j2]: This is a change from the previous default

File: go-controller/pkg/config/config.go [hunk 18]
```diff
@@ -846,7 +846,6 @@ func initConfigWithPath(ctx *cli.Context, exec kexec.Interface, saPath string, d
 			return "", fmt.Errorf("failed to parse config file %s: %v", f.Name(), err)
 		}
 		logrus.Infof("Parsed config file %s", f.Name())
-		logrus.Infof("Parsed config: %+v", cfg)
```
Comment (danwinship) [inline go-controller/pkg/config/config.go]: why was this removed?

File: dist/images/README_IMAGE.md (hunk 2 above)
Comment (dcbw) [reply dist/images/README_IMAGE.md]: @danwinship I think the one big reason we wanted a ConfigMap was to reduce the size/churn of ovnkube.sh. With a CM we can set many options, including new ones, from the DaemonSet without having to rev ovnkube.sh for every small change. And honestly I'd rather kill the environment variables so that we only have 2 ways of delivering config (CLI and config file) rather than 3 (CLI, file, env). ConfigMap also allows us to dynamically change config in the future (not that we'd use it now or soon).

Comment (danwinship) [reply dist/images/README_IMAGE.md]: No, I mean, there are two ConfigMaps here; `ovnkube-config` gets passed as a file to `/usr/bin/ovnkube` and that part totally makes sense. But `ovn-config` is the stuff that `ovnkube.sh` itself uses directly, and it just copies all of its contents into environment variables. Except that we already had code to deal with that in v3 by having the pod spec copy the configmap into environment variables, and it's not clear what the v4 code gains us there.

Now if you wanted to have `ovnkube.sh` parse the `cluster-subnets`, `service-cidr`, `ovn-config-namespace`, and `apiserver` out of `ovnkube-config`, and then get rid of the `ovn-config` ConfigMap entirely, that would make sense.

**Comment by girishmg** [comment]
@dcbw @danwinship 

There is some overlap with the PR #820. Furthermore, we also need to make sure that this PR doesn't preclude us from supporting multiple OVN interfaces into the POD.

As I see it, there will be following configuration knobs:

1. Per-container environment variables that resides in
   Container’s specification
   - K8S_NODE
   - OVN_KUBERNETES_NAMESPACE
   - respective OVN_LOG_* setting

2. OVN CNI specific configuration (PR #820)
   - NET_CIDR for OVN logical topology
   - MTU for OVN logical topology
   - OVN ENCAP IP

3. ovnkube daemon specific configuration
    - SSL certificates
    - OVN DB IPs
    - Gateway configuration 
    - logging

**Comment by pecameron** [comment]
@dcbw @danwinship On config maps. onvkube-config is the config file for ovnkube. Whatever is in there needs to be parsed and accepted by ovnkube. ovn-config is everything else that we used to have in environment variables. We have also added secret and ca files. So we now have environment variables, mounted files and two configmaps that establish the API between cno and ovn. If anything changes in any of these places we need to evaluate whether to change the API version. It was much easier when we just had environment variables. I am 100% in favor of just using environment variables.

When the desire is to reduce the size and complexity of ovnkube.sh adding parsing and supporting multiple ways of passing data seem to move in the wrong direction. 

The API is alive and changing. We are going to have some churn in at least the ovnkube-config and ovn-config configmaps when ssl is added.

Further, whenever any of these items change we need to make sure things remain compatible between the current and the changed. All permutations need to be tested. It is not good enough to hack ovn and cno and test the changes together. Each PR must always pass CI in its repo since there will likely be multiple PR is each repo at any point in time. It is very possible for a PR to pass and get merged in OVN and the resulting new image breaks CI in CNO. Eventually the CNO PR will get reviewed, merged and CI will pass again. In the meantime testing for all other cno PRs is broken. This is not a problem for upstream since everything is in the same repo and a PR can fix everything.

**Comment by pecameron** [comment]
@danwinship dist/images/ovnkube.sh line 701
wait_for_event process_ready ovn-controller now polls for the ovn-controller pid. the sleep 4 was left over from 3.11 bring up days.

Bear in mind that the wait may not wait long enough since the ovn-controller needs to come up and write the cni file before moving on. This race doesn't seem to come out wrong. Also, fixing races with timeouts is never a good idea.

**Comment by pecameron** [comment]
@danwinship dist/templates/ovn-setup.yaml.j2 line 54 mode=local
A default is exactly that, a default. We have been using mode=local for a while. It has been specified in various places over time.

**Comment by pecameron** [comment]
@girishmg @dcbw @danwinship comments on PR820 
2. OVN CNI specific configuration
There needs to be some way to specify this on a per network basis. We also need to specify the interactions between the networks. For example, I can have a storage network that is totally separate from the cluster network. Or I can have a data network that can gateway into the cluster network. 
The current ovn-config configmap needs to be changed.

3. ovnkube daemon specific configuration
We use utilities in the script that access the sbdb in addition to ovnkube. The ssl config needs to be present for this.

As an aside, we need to eliminate the need for kubectl in the ovnkube.sh script.

**Comment by danwinship** [comment]
> It is very possible for a PR to pass and get merged in OVN and the resulting new image breaks CI in CNO. Eventually the CNO PR will get reviewed, merged and CI will pass again. In the meantime testing for all other cno PRs is broken.

As far as I know that hasn't happened yet, and once the `e2e-aws-ovn-kubernetes` jobs are fully functional, it won't be possible at all; you will not be able to merge a PR to `openshift/ovn-kubernetes` that is incompatible with the current state of `openshift/cluster-network-operator`, and vice versa. In some cases, this means that changes will need to be done in multiple steps (when changing ovn-kubernetes from "old thing" to "new thing", there will need to be an intermediate stage where either `openshift/ovn-kubernetes` or `openshift/cluster-network-operator` supports both the old and the new thing; like how this PR still supports v3 while adding support for v4).

**Comment by danwinship** [comment]
> A default is exactly that, a default. We have been using mode=local for a while. It has been specified in various places over time.

*OpenShift* uses mode=local, but the default in `ovn-org/ovn-kubernetes` was mode=shared. I assume there was some reason why someone wanted the upstream default to be "shared", so I pointed out the change in your PR. If no one cares then we can change the default to "local"

**Comment by dcbw** [comment]
> There is some overlap with the PR #820. Furthermore, we also need to make sure that this PR doesn't preclude us from supporting multiple OVN interfaces into the POD.
> 
> As I see it, there will be following configuration knobs:
> 
>     1. Per-container environment variables that resides in
>        Container’s specification
>        
>        * K8S_NODE
>        * OVN_KUBERNETES_NAMESPACE
>        * respective OVN_LOG_* setting

@girishmg I don't think any of these are per-container? At least the OVN_KUBERNETES_NAMESPACE isn't, that's daemon-wide information unless we're going to start running multiple parallel ovn-kubernetes daemonsets in the cluster. Instead of that, I thought we were going to have one daemonset that handled the additional networks?

Perhaps I misunderstand how we are using these though.

>     2. OVN CNI specific configuration (PR #820)
>        
>        * NET_CIDR for OVN logical topology
>        * MTU for OVN logical topology
>        * OVN ENCAP IP

Yes, these would be "network" specific. But I thought we could use these for the first "default" network, and then the other networks have their own specific config? eg these would come from other Kube API objects.

>     3. ovnkube daemon specific configuration
>        
>        * SSL certificates
>        * OVN DB IPs
>        * Gateway configuration
>        * logging

Yep, clearly daemonset-wide config that I think can be passed in the ConfigMap.

**Comment by pecameron** [comment]
@danwinship @dcbw New round of changes. PTAL

File: dist/images/Dockerfile.fedora [hunk 19]
```diff
@@ -20,7 +20,7 @@ RUN INSTALL_PKGS=" \
 	PyYAML bind-utils procps-ng openssl numactl-libs firewalld-filesystem \
 	libpcap hostname kubernetes-client \
         ovn ovn-central ovn-host \
-	iptables iproute strace socat \
+	iptables-nft iproute strace socat \
```
[the remark this answers is not in the document]
Comment (pecameron) [reply dist/images/Dockerfile.fedora]: Dropped from PR

File: dist/images/ovnkube.sh (hunk 15 above)
Comment (pecameron) [reply dist/images/ovnkube.sh]: Early code (3.11) needed this to come up. Later added "wait_for_event process_ready ovn-controller" forgot to remove sleep

File: dist/templates/ovn-setup.yaml.j2 (hunk 17 above)
Comment (pecameron) [reply dist/templates/ovn-setup.yaml.j2]: Gateway code was added and this was set to local. The actual default stays as "shared".

**Comment by pecameron** [comment]
No longer using ovnkube.sh
